internal/internal_workflow.go

```diff
@@ -1046,23 +1084,26 @@ func (s *coroutineState) stackTrace() string {
 	return <-stackCh
 }
 
+func (s *coroutineState) run(ctx Context, f func(ctx Context)) {
+	defer runtime.KeepAlive(&s) // keep receiver argument alive for getCoroStackTrace
```

Comment:
Can you help me understand this a bit? `s` is passed to `getCoroStackTrace` so doesn't that imply liveness? (also, not sure it's worth breaking this call out to a separate method makes since unless you are saying that the pointer is unstable)

Reply:
Sure! The liveness problem doesn't concern the `coroutineState` itself, but the pointer *value* `s` as an argument to this function. Without the `KeepAlive` call, the last use of `s` is at `s.initialYield(...)` (deferred calls evaluate their arguments [at the point of the defer statement](https://go.dev/ref/spec#Defer_statements)). Hence, when `f` is called, Go considers `s` dead **and doesn't print its value anymore in the goroutine's stack trace**. The search then can't find the stack we want.

`runtime.KeepAlive` is a special function in the sense that it's a no-op, but marks its argument as "used" for Go's liveness analysis. So replacing it with some regular Go statement runs the risk of the compiler optimizing the statement out.

To make sure `s` is alive throughout the entire function (including deferred calls!), `runtime.KeepAlive(s)` needs to be the very last statement. This means it must be deferred first, but since deferred calls evaluate their arguments eagerly, the value that is being kept alive is a copy of `s`. By instead keeping `&s` alive, `s` must also be kept alive since `&s` points to it.

Reply:
N.B. the same concerns apply to any function argument we might want to use to identify the coroutine. This is not specific to `s` being a pointer itself.